File: surface/widgets/nativescript/components/widgets/concepts/llm-agent/TaskPlanList.ts
import { StackLayout, Label, ScrollView, FlexboxLayout, Progress } from '@nativescript/core';

/* ---------------------------------------------------------------------------
 * TaskPlanList state machine
 * ------------------------------------------------------------------------- */

export type TaskPlanListState = 'idle' | 'taskSelected' | 'reordering';
export type TaskPlanListEvent =
  | { type: 'EXPAND_TASK'; id?: string }
  | { type: 'COLLAPSE_TASK'; id?: string }
  | { type: 'SELECT_TASK'; id?: string }
  | { type: 'DRAG_START' }
  | { type: 'DESELECT' }
  | { type: 'DROP' }
  | { type: 'CANCEL_DRAG' };

export function taskPlanListReducer(state: TaskPlanListState, event: TaskPlanListEvent): TaskPlanListState {
  switch (state) {
    case 'idle':
      if (event.type === 'EXPAND_TASK') return 'idle';
      if (event.type === 'COLLAPSE_TASK') return 'idle';
      if (event.type === 'SELECT_TASK') return 'taskSelected';
      if (event.type === 'DRAG_START') return 'reordering';
      return state;
    case 'taskSelected':
      if (event.type === 'DESELECT') return 'idle';
      if (event.type === 'SELECT_TASK') return 'taskSelected';
      return state;
    case 'reordering':
      if (event.type === 'DROP') return 'idle';
      if (event.type === 'CANCEL_DRAG') return 'idle';
      return state;
    default:
      return state;
  }
}

/* ---------------------------------------------------------------------------
 * Types
 * ------------------------------------------------------------------------- */

export type TaskStatus = 'pending' | 'active' | 'complete' | 'failed' | 'skipped';

export interface Task {
  id: string;
  label: string;
  status: TaskStatus;
  result?: string;
  subtasks?: Task[];
}

export interface TaskPlanListProps {
  tasks: Task[];
  goalLabel: string;
  progress: number;
  showProgress?: boolean;
  allowReorder?: boolean;
  expandedTasks?: string[];
  onReorder?: (tasks: Task[]) => void;
}

/* ---------------------------------------------------------------------------
 * Helpers
 * ------------------------------------------------------------------------- */

const STATUS_ICONS: Record<TaskStatus, string> = {
  pending: '\u25CB',
  active: '\u25CF',
  complete: '\u2713',
  failed: '\u2717',
  skipped: '\u2298',
};

const STATUS_LABELS: Record<TaskStatus, string> = {
  pending: 'Pending',
  active: 'Active',
  complete: 'Complete',
  failed: 'Failed',
  skipped: 'Skipped',
};

function countTasks(tasks: Task[]): { complete: number; total: number } {
  let complete = 0;
  let total = 0;
  for (const t of tasks) {
    total += 1;
    if (t.status === 'complete') complete += 1;
    if (t.subtasks) {
      const sub = countTasks(t.subtasks);
      complete += sub.complete;
      total += sub.total;
    }
  }
  return { complete, total };
}

/* ---------------------------------------------------------------------------
 * Widget
 * ------------------------------------------------------------------------- */

export function createTaskPlanList(props: TaskPlanListProps): { view: StackLayout; dispose: () => void } {
  const {
    tasks,
    goalLabel,
    progress,
    showProgress = true,
    expandedTasks: expandedTasksProp = [],
  } = props;

  let widgetState: TaskPlanListState = 'idle';
  let expandedSet = new Set<string>(expandedTasksProp);
  let selectedId: string | null = null;
  const disposers: (() => void)[] = [];

  function send(event: TaskPlanListEvent) {
    widgetState = taskPlanListReducer(widgetState, event);
  }

  const root = new StackLayout();
  root.className = 'task-plan-list';
  root.automationText = `Task plan: ${goalLabel}`;

  // Goal header
  const goalHeader = new Label();
  goalHeader.className = 'task-plan-list-goal';
  goalHeader.text = goalLabel;
  root.addChild(goalHeader);

  // Progress bar
  if (showProgress) {
    const { complete: completedCount, total: totalCount } = countTasks(tasks);
    const progressPct = Math.min(Math.max(progress, 0), 100);

    const progressBar = new Progress();
    progressBar.className = 'task-plan-list-progress';
    progressBar.value = progressPct;
    progressBar.maxValue = 100;
    progressBar.automationText = `${completedCount} of ${totalCount} tasks complete`;
    root.addChild(progressBar);

    const progressText = new Label();
    progressText.className = 'task-plan-list-progress-text';
    progressText.text = `${completedCount} of ${totalCount} tasks complete`;
    root.addChild(progressText);
  }

  // Task list area
  const scrollView = new ScrollView();
  const taskContainer = new StackLayout();
  taskContainer.className = 'task-plan-list-tasks';
  scrollView.content = taskContainer;
  root.addChild(scrollView);

  function renderTask(task: Task, depth: number, container: StackLayout) {
    const hasSubtasks = !!(task.subtasks && task.subtasks.length > 0);
    const isExpanded = expandedSet.has(task.id);
    const isSelected = selectedId === task.id;

    const taskItem = new FlexboxLayout();
    taskItem.className = isSelected ? 'task-plan-list-item-selected' : 'task-plan-list-item';
    taskItem.flexDirection = 'row' as any;
    taskItem.alignItems = 'center' as any;
    taskItem.paddingLeft = depth * 24;
    taskItem.automationText = `${task.label} \u2014 ${STATUS_LABELS[task.status]}`;

    const statusIcon = new Label();
    statusIcon.className = `task-plan-list-status task-plan-list-status-${task.status}`;
    statusIcon.text = STATUS_ICONS[task.status];
    taskItem.addChild(statusIcon);

    const taskLabel = new Label();
    taskLabel.className = 'task-plan-list-task-label';
    taskLabel.text = task.label;
    taskItem.addChild(taskLabel);

    if (hasSubtasks) {
      const expandIcon = new Label();
      expandIcon.className = 'task-plan-list-expand-icon';
      expandIcon.text = isExpanded ? '\u25BC' : '\u25B6';
      taskItem.addChild(expandIcon);
    }

    const tapHandler = () => {
      if (selectedId === task.id) {
        selectedId = null;
        send({ type: 'DESELECT' });
      } else {
        selectedId = task.id;
        send({ type: 'SELECT_TASK', id: task.id });
      }

      if (hasSubtasks) {
        if (expandedSet.has(task.id)) {
          expandedSet.delete(task.id);
          send({ type: 'COLLAPSE_TASK', id: task.id });
        } else {
          expandedSet.add(task.id);
          send({ type: 'EXPAND_TASK', id: task.id });
        }
      }
      rebuildTasks();
    };
    taskItem.on('tap', tapHandler);

    container.addChild(taskItem);

    if (isExpanded && task.result) {
      const resultLabel = new Label();
      resultLabel.className = 'task-plan-list-result';
      resultLabel.text = task.result;
      resultLabel.textWrap = true;
      resultLabel.paddingLeft = (depth + 1) * 24;
      container.addChild(resultLabel);
    }

    if (isExpanded && hasSubtasks) {
      for (const sub of task.subtasks!) {
        renderTask(sub, depth + 1, container);
      }
    }
  }

  function rebuildTasks() {
    taskContainer.removeChildren();
    for (const task of tasks) {
      renderTask(task, 0, taskContainer);
    }
  }

  rebuildTasks();

  return {
    view: root,
    dispose() {
      disposers.forEach((d) => d());
    },
  };
}

export default createTaskPlanList;

File: surface/widgets/swiftui/components/widgets/concepts/llm-agent/TaskPlanList.swift
import SwiftUI

// MARK: - Types

enum TaskStatus: String, CaseIterable { case pending, active, complete, failed, skipped }

struct PlanTask: Identifiable {
    let id: String
    let label: String
    let status: TaskStatus
    var result: String?
    var subtasks: [PlanTask]?
}

// MARK: - State Machine

enum TaskPlanListState { case idle, taskSelected, reordering }
enum TaskPlanListEvent {
    case expandTask(id: String), collapseTask(id: String), selectTask(id: String)
    case dragStart, deselect, drop, cancelDrag
}

func taskPlanListReduce(state: TaskPlanListState, event: TaskPlanListEvent) -> TaskPlanListState {
    switch state {
    case .idle:
        switch event {
        case .expandTask, .collapseTask: return .idle
        case .selectTask: return .taskSelected
        case .dragStart: return .reordering
        default: return state
        }
    case .taskSelected:
        switch event {
        case .deselect: return .idle
        case .selectTask: return .taskSelected
        default: return state
        }
    case .reordering:
        switch event {
        case .drop, .cancelDrag: return .idle
        default: return state
        }
    }
}

private let statusIcons: [TaskStatus: String] = [
    .pending: "\u{25CB}", .active: "\u{25CF}", .complete: "\u{2713}", .failed: "\u{2717}", .skipped: "\u{2298}"
]

private func countTasks(_ tasks: [PlanTask]) -> (complete: Int, total: Int) {
    var c = 0, t = 0
    for task in tasks {
        t += 1; if task.status == .complete { c += 1 }
        if let subs = task.subtasks { let s = countTasks(subs); c += s.complete; t += s.total }
    }
    return (c, t)
}

// MARK: - View

struct TaskPlanListView: View {
    let tasks: [PlanTask]
    let goalLabel: String
    let progress: Double
    var showProgress: Bool = true
    var allowReorder: Bool = true
    var onReorder: (([PlanTask]) -> Void)?

    @State private var widgetState: TaskPlanListState = .idle
    @State private var expandedSet: Set<String> = []
    @State private var selectedId: String? = nil

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            Text(goalLabel).font(.headline).padding(.horizontal, 12).padding(.vertical, 8)

            if showProgress {
                let counts = countTasks(tasks)
                VStack(spacing: 2) {
                    GeometryReader { geo in
                        ZStack(alignment: .leading) {
                            RoundedRectangle(cornerRadius: 3).fill(Color.gray.opacity(0.2))
                            RoundedRectangle(cornerRadius: 3).fill(Color.blue)
                                .frame(width: geo.size.width * CGFloat(min(max(progress, 0), 100) / 100))
                        }
                    }.frame(height: 6)
                    Text("\(counts.complete) of \(counts.total) tasks complete")
                        .font(.caption2).foregroundColor(.secondary)
                }
                .padding(.horizontal, 12).padding(.bottom, 8)
                .accessibilityLabel("\(counts.complete) of \(counts.total) tasks complete")
            }

            Divider()

            ScrollView {
                LazyVStack(alignment: .leading, spacing: 0) {
                    ForEach(tasks) { task in
                        taskRow(task, depth: 0)
                    }
                }
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Task plan: \(goalLabel)")
    }

    @ViewBuilder
    private func taskRow(_ task: PlanTask, depth: Int) -> some View {
        let isExpanded = expandedSet.contains(task.id)
        let isSelected = selectedId == task.id
        let hasSubtasks = !(task.subtasks ?? []).isEmpty

        VStack(alignment: .leading, spacing: 0) {
            HStack(spacing: 8) {
                Text(statusIcons[task.status] ?? "\u{25CB}").font(.system(size: 14))
                    .foregroundColor(task.status == .active ? .blue : (task.status == .failed ? .red : .primary))
                Text(task.label).lineLimit(1)
                Spacer()
                if hasSubtasks {
                    Button(action: { toggleExpand(task.id) }) {
                        Text(isExpanded ? "\u{25BC}" : "\u{25B6}").font(.system(size: 10))
                    }.buttonStyle(.plain)
                }
            }
            .padding(.leading, CGFloat(depth * 20))
            .padding(.horizontal, 12).padding(.vertical, 6)
            .background(isSelected ? Color.blue.opacity(0.1) : Color.clear)
            .contentShape(Rectangle())
            .onTapGesture {
                if selectedId == task.id {
                    selectedId = nil
                    widgetState = taskPlanListReduce(state: widgetState, event: .deselect)
                } else {
                    selectedId = task.id
                    widgetState = taskPlanListReduce(state: widgetState, event: .selectTask(id: task.id))
                }
            }
            .accessibilityLabel("\(task.label) \u{2014} \(task.status.rawValue.capitalized)")

            if isExpanded, let result = task.result {
                Text(result).font(.system(size: 13)).foregroundColor(.secondary)
                    .padding(.leading, CGFloat((depth + 1) * 20)).padding(.horizontal, 12).padding(.vertical, 4)
            }

            if isExpanded, let subs = task.subtasks {
                ForEach(subs) { sub in taskRow(sub, depth: depth + 1) }
            }
        }
    }

    private func toggleExpand(_ id: String) {
        if expandedSet.contains(id) { expandedSet.remove(id) } else { expandedSet.insert(id) }
    }
}
